cuDSS/simple_matrix_market/matrix_market_reader.h

```diff
@@ -0,0 +1,250 @@
+        fprintf(stdout, "WARNING: mview is full, but only lower or upper elements found\n");
+    }
+
+    //Initialize with 0's
+    std::fill(*csr_offsets_h, *csr_offsets_h + (n + 1), 0);
+
+    int current_idx = 0;
+
+    for (auto& [i, j, val] : entries) {
+        if (i >= n || i < 0) {
+            fprintf(stderr, "ERROR: Invalid row index %d\n", i);
+            return MtxReaderErrorOutOfBoundRowIndex;
+        }
+        if (j >= n || j < 0) {
+            fprintf(stderr, "ERROR: Invalid col index %d\n", j);
+            return MtxReaderErrorOfBoundColIndex;
+        }
+        (*csr_offsets_h)[i + 1]++;
+        (*csr_columns_h)[current_idx] = j;
+        (*csr_values_h)[current_idx] = val;
+        current_idx++;
+    }
+
+    // Prefix sum to compute the offsets
+    for (int i = 0; i < n; ++i) 
+        (*csr_offsets_h)[i + 1] += (*csr_offsets_h)[i];
+        
+
+    // Detect empty rows
+    for (int i = 0; i < n; ++i) {
+        if ((*csr_offsets_h)[i] == (*csr_offsets_h)[i + 1]) {
+            printf("Warning: Row %d is empty\n", i);
+        }
+    }
+
+    printf("MATRIX READER: Completed with %d nonzeros\n", nnz);
+    return 0;
+}
+
+template <typename value_type>
+int rhs_reader(std::string filename, int& n, value_type** b_values_h) {
+    std::ifstream file(filename);
+
+    if (!file.is_open()) {
+        fprintf(stderr, "Error: Could not open file %s\n", filename.c_str());
+        return EXIT_FAILURE;
+    }
+
+    std::string line;
+    bool foundSize = false;
+    bool foundHeader = false;
+    int row = 0;
+    while (std::getline(file, line)) {
+        //Get line
+        std::istringstream lineData(line);
+        
+        //skip empty lines
+        if (line.empty()) 
+            continue;
+
+        //skip comments
+        if (foundHeader && line[0]=='%') 
+            continue;
+
+        //Get header
+        if (!foundHeader && line.substr(0, 14) == "%%MatrixMarket"){
+            foundHeader = true;
+                
+            // Parse header: %%MatrixMarket matrix coordinate real general
+            std::string marker, object, format, value_type_header, symmetry;
+            
+            lineData >> marker >> object >> format >> value_type_header >> symmetry;
+            
+            // Check if it's matrix coordinate real
+            if (object != "matrix" || format != "array" || value_type_header != "real") {
+                fprintf(stderr, "ERROR: Expected 'matrix array real for the rhs file header\n");
+                fprintf(stderr, "Got: object=%s format=%s value_type=%s\n", 
+                        object.c_str(), format.c_str(), value_type_header.c_str());
+                    return MtxReaderErrorInvalidFormatInHeader;
+            }
+            continue;
+        }
+
+        if (!foundSize) {
+            int num_rhs;
+            lineData >> n >> num_rhs;
+            if (num_rhs != 1) {
+                fprintf(stderr, "Only one RHS is supported, %d found.", num_rhs);
```

Comment:
nit: Missing the newline in the end:

```suggestion
                fprintf(stderr, "Only one RHS is supported, %d found.\n", num_rhs);
```